File: prisme_core/routes/vecteurs.py
"""Routes de la recherche sémantique : état, vectorisation, diagnostic."""
from flask import Blueprint, jsonify, request

from ..config import rd_cfg, wr_cfg
from ..vecteurs import contrat, magasin, vectorisation

bp = Blueprint("vecteurs", __name__)
PREFIX = "/api/vecteurs"


@bp.route(PREFIX + "/etat", methods=["GET"])
def etat():
    """Ce qui est actif, où, et avec quoi. Consulté en permanence par l'interface."""
    return jsonify({**vectorisation.etat(), "fournisseurs": contrat.noms()})


@bp.route(PREFIX + "/tester", methods=["POST"])
def tester():
    """Sonde le fournisseur sans rien écrire : dit s'il répond et en quelle dimension."""
    d = request.get_json(silent=True) or {}
    cfg = {**rd_cfg(), **{k: v for k, v in d.items() if k.startswith("emb_")}, "emb_actif": True}
    try:
        f = contrat.construire(cfg.get("emb_fournisseur") or "api", cfg)
        ok, raison = f.disponible()
        if not ok:
            return jsonify({"ok": False, "raison": raison}), 200
        return jsonify({"ok": True, "modele": f.modele(), "dimension": f.dimension(),
                        "distant": bool(f.distant), "signature": f.signature()})
    except contrat.VecteurIndisponible as e:
        return jsonify({"ok": False, "raison": str(e)}), 200


@bp.route(PREFIX + "/vectoriser", methods=["POST"])
def vectoriser():
    d = request.get_json(silent=True) or {}
    if d.get("fond"):
        return jsonify({"lance": vectorisation.lancer_en_fond(), **vectorisation.ETAT})
    resultat = vectorisation.vectoriser()
    return (jsonify(resultat), 400) if resultat.get("error") else jsonify(resultat)


@bp.route(PREFIX + "/progression", methods=["GET"])
def progression():
    return jsonify(dict(vectorisation.ETAT))


@bp.route(PREFIX + "/vider", methods=["POST"])
def vider():
    """Efface les vecteurs. L'index et les notes ne sont pas touchés."""
    m = magasin.magasin_pour()
    avant = m.compte()
    m.vider()
    return jsonify({"ok": True, "supprimes": avant})


@bp.route(PREFIX + "/config", methods=["POST"])
def config():
    """Enregistre le paramétrage des embeddings, sans toucher à celui du chat."""
    d = request.get_json(silent=True) or {}
    champs = {k: d[k] for k in ("emb_actif", "emb_fournisseur", "emb_base_url",
                                "emb_modele", "emb_api_key", "emb_seuil") if k in d}
    if "emb_seuil" in champs:
        try:
            champs["emb_seuil"] = min(0.95, max(0.0, float(champs["emb_seuil"])))
        except (TypeError, ValueError):
            return jsonify({"error": "Seuil invalide : un nombre entre 0 et 0,95"}), 400
    if "emb_fournisseur" in champs and champs["emb_fournisseur"] not in contrat.noms():
        return jsonify({"error": "Fournisseur inconnu : %s" % champs["emb_fournisseur"]}), 400
    if "emb_actif" in champs:
        champs["emb_actif"] = bool(champs["emb_actif"])
    wr_cfg(champs)
    return jsonify(vectorisation.etat())

File: prisme_core/web/js/vecteurs.js
// ══════════════════════════════════════════════════
//  RECHERCHE SÉMANTIQUE — PARAMÉTRAGE ET VECTORISATION
// ══════════════════════════════════════════════════
// Désactivée par défaut : vectoriser, c'est parfois envoyer le texte de ses notes à
// un tiers. Le mode actif reste visible en permanence, et ce qui sort de la machine
// est dit sans détour (décision 0019).
var VEC = {etat: null, timer: null};

async function vecCharger(){
  clearTimeout(VEC.timer);
  var d = await fetch('/api/vecteurs/etat').then(function(r){ return r.json(); });
  VEC.etat = d;
  $('vec-actif').checked = !!d.actif;
  $('vec-fournisseur').innerHTML = (d.fournisseurs || []).map(function(f){
    return '<option value="' + esc(f) + '">' + esc(vecLibelle(f)) + '</option>';
  }).join('');
  $('vec-fournisseur').value = d.fournisseur || 'api';
  $('vec-modele').value = d.modele || '';
  vecRendreEtat(d);
  if(d.progression && d.progression.en_cours) VEC.timer = setTimeout(vecCharger, 800);
}

function vecLibelle(f){
  return {api: 'API compatible OpenAI (/embeddings)', ollama: 'Ollama (local)',
          onnx: 'ONNX local (plugin)'}[f] || f;
}

function vecRendreEtat(d){
  var e = $('vec-etat');
  var p = d.progression || {};
  if(p.en_cours){
    e.className = 'vec-etat en-cours';
    e.textContent = '⏳ Vectorisation… ' + p.faits + ' / ' + p.total + ' segment(s)';
    return;
  }
  if(!d.actif){
    e.className = 'vec-etat off';
    e.textContent = '○ Désactivée — recherche par mots seule. '
      + (d.vecteurs ? d.vecteurs + ' vecteur(s) conservé(s).' : '');
    return;
  }
  if(!d.disponible){
    e.className = 'vec-etat ko';
    e.textContent = '⚠ ' + (d.raison || 'Fournisseur indisponible') + ' — repli sur la recherche par mots.';
    return;
  }
  e.className = 'vec-etat ok';
  e.textContent = (d.distant ? '⚠ ACTIVE, MODE DISTANT — le texte de vos notes est envoyé à '
                             : '● Active, en local — rien ne sort de votre machine. ')
    + (d.distant ? (d.signature_attendue || '').split(':')[0] + ' pour être vectorisé. ' : '')
    + d.vecteurs + ' segment(s) vectorisé(s)'
    + (d.signature ? ' · ' + esc(d.signature) : '');
  if(p.erreur) e.textContent += ' · dernière erreur : ' + p.erreur;
}

async function vecTester(){
  var btn = $('vec-tester'); btn.disabled = true; btn.textContent = '⏳…';
  var d = await post('/api/vecteurs/tester', vecChamps());
  btn.disabled = false; btn.textContent = 'Tester';
  if(!d.ok){ toast('⚠ ' + (d.raison || d.error), 6000); return; }
  toast('✓ ' + d.modele + ' · ' + d.dimension + ' dimensions · '
        + (d.distant ? 'service distant' : 'local'), 5000);
}

function vecChamps(){
  var c = {emb_fournisseur: $('vec-fournisseur').value, emb_modele: $('vec-modele').value.trim(),
           emb_base_url: $('vec-url').value.trim()};
  var k = $('vec-cle').value;
  if(k) c.emb_api_key = k;
  return c;
}

async function vecEnregistrer(){
  var c = vecChamps();
  c.emb_actif = $('vec-actif').checked;
  if(c.emb_actif && !c.emb_modele){ toast('⚠ Indiquez le modèle d\'embeddings'); return; }
  var d = await post('/api/vecteurs/config', c);
  if(d.error){ toast('⚠ ' + d.error); return; }
  $('vec-cle').value = '';
  toast('✓ Recherche sémantique enregistrée');
  vecCharger();
}

async function vecVectoriser(){
  if(!VEC.etat || !VEC.etat.actif){ toast('Activez d\'abord la recherche sémantique'); return; }
  if(VEC.etat.distant){
    if(!await confirmer({titre: 'Envoi à un service distant', ok: 'Vectoriser', danger: true,
        message: 'Le texte de toutes vos notes va être envoyé au service configuré pour y être '
          + 'vectorisé. Rien n\'est envoyé pour les notes déjà vectorisées. Continuer ?'})) return;
  }
  var d = await post('/api/vecteurs/vectoriser', {fond: true});
  if(d.error){ toast('⚠ ' + d.error); return; }
  toast(d.lance ? 'Vectorisation lancée — vous pouvez continuer à travailler'
                : 'Une vectorisation est déjà en cours');
  vecCharger();
}

async function vecVider(){
  if(!await confirmer({titre: 'Effacer les vecteurs', ok: 'Effacer', danger: true,
      message: 'Les vecteurs seront supprimés. Vos notes et l\'index ne sont pas touchés, '
        + 'mais il faudra tout revectoriser — ce qui peut être long, voire payant.'})) return;
  var d = await post('/api/vecteurs/vider', {});
  toast(d.supprimes + ' vecteur(s) effacé(s)');
  vecCharger();
}
